tests/e2e/tests/upgrade/upgrade_test.go

```diff
@@ -0,0 +1,358 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e2e
+
+import (
+	"errors"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"testing"
+	"time"
+	// TODO(nmittler): Remove this
+	_ "github.com/golang/glog"
+	multierror "github.com/hashicorp/go-multierror"
+
+	"istio.io/istio/pkg/log"
+	"istio.io/istio/tests/e2e/framework"
+	"istio.io/istio/tests/util"
+)
+
+const (
+	u1                       = "normal-user"
+	u2                       = "test-user"
+	bookinfoYaml             = "samples/bookinfo/kube/bookinfo.yaml"
+	bookinfoRatingsv2Yaml    = "samples/bookinfo/kube/bookinfo-ratings-v2.yaml"
+	bookinfoRatingsMysqlYaml = "samples/bookinfo/kube/bookinfo-ratings-v2-mysql.yaml"
+	bookinfoDbYaml           = "samples/bookinfo/kube/bookinfo-db.yaml"
+	bookinfoMysqlYaml        = "samples/bookinfo/kube/bookinfo-mysql.yaml"
+	modelDir                 = "tests/apps/bookinfo/output"
+	rulesDir                 = "samples/bookinfo/kube"
+	allRule                  = "route-rule-all-v1.yaml"
+	testRule                 = "route-rule-reviews-test-v2.yaml"
+)
+
+var (
+	tc              *testConfig
+	testRetryTimes  = 5
+	defaultRules    = []string{allRule, testRule}
+	flagBaseVersion = flag.String("base_version", "0.4.0", "Base version to use for upgrade.")
+)
+
+type testConfig struct {
+	*framework.CommonConfig
+	gateway  string
+	rulesDir string
+}
+
+func (t *testConfig) Setup() error {
+	t.gateway = "http://" + tc.Kube.Ingress
+	//generate rule yaml files, replace "jason" with actual user
+	for _, rule := range defaultRules {
+		src := util.GetResourcePath(filepath.Join(rulesDir, rule))
+		dest := filepath.Join(t.rulesDir, rule)
+		ori, err := ioutil.ReadFile(src)
+		if err != nil {
+			log.Errorf("Failed to read original rule file %s", src)
+			return err
+		}
+		content := string(ori)
+		content = strings.Replace(content, "jason", u2, -1)
+		err = ioutil.WriteFile(dest, []byte(content), 0600)
+		if err != nil {
+			log.Errorf("Failed to write into new rule file %s", dest)
+			return err
+		}
+
+	}
+
+	if !util.CheckPodsRunning(tc.Kube.Namespace) {
+		return fmt.Errorf("can't get all pods running")
+	}
+
+	return setUpDefaultRouting()
+}
+func getWithCookie(url string, cookies []http.Cookie) (*http.Response, error) {
+	// Declare http client
+	client := &http.Client{}
+
+	// Declare HTTP Method and Url
+	req, err := http.NewRequest(http.MethodGet, url, nil)
+	if err != nil {
+		return nil, err
+	}
+	for _, c := range cookies {
+		// Set cookie
+		req.AddCookie(&c)
+	}
+	return client.Do(req)
+}
+
+func closeResponseBody(r *http.Response) {
```

Comment:
I see many duplicated functions from bookinfo test, such as inspect, closeRespoinseBody, deleteRules, applyRules, etc. Perhaps it is better to put them in a shared utility.

Reply:
Those are specific for bookinfo app, so maybe not suitable for general test utils.